src/app/trainees/trainees.component.html

```diff
@@ -5,3 +5,7 @@
 <div class="p-15">
   <router-outlet></router-outlet>
 </div>
+
```

Comment:
Tying component instances with routing may have an adverse effect with events. e.g. ngOnInit, etc will fire at separate instances out of sync if event based data. Just thinking aloud if clearFilters, pageChanges etc affect routerNagivationEnd, then data will be out of sync. If components are all loaded in tranee-list @inputs and outputs may keep them in sync? Maybe? trainee lists may keep subscribing to parameter changes (this.route.queryParamMap.subscribe || this.route.queryParams.subscribe) and fire actions appropriately? I will download this branch and see events sequences with sorting and page changing to see ... once again just thinking aloud

Reply:
i dont think i quite understand. the route change is basically done to gracefully update the url which doesn't reinstantiate the component or fire ngOnInit either. 

the queryParams check via url snapshot is to cater for the scenario if `/trainees` route is viewed without any parameters it lets the store know how to update appropriately and subsequently make the api call with the correct default parameters

Reply:
yeah i just checked, we're firing quite a bit in different events.

![Screenshot 2020-04-07 at 11 35 57](https://user-images.githubusercontent.com/13540819/78660687-711f7980-78c5-11ea-9b4a-05419be16ce1.png)

Reply:
i'll pull-down and check again. I suspect we're talking about different things here

Reply:
> i dont think i quite understand. the route change is basically done to gracefully update the url which doesn't reinstantiate the component or fire ngOnInit either.
> 
> the queryParams check via url snapshot is to cater for the scenario if `/trainees` route is viewed without any parameters it lets the store know how to update appropriately and subsequently make the api call with the correct default parameters

In response: ngOnint always fires on route change if your component loads in router-outlet

Reply:
the BE work for pagination hasn't completed yet so the api response is sending back thousands of records and the ui therefore has a super long almost never ending scroll. i would imagine the paginator is out of view but not to worry as I've got another task to integrate the pagination with the BE once uzair is done. 

ngOnInit doesn't fire when navigating to the same route which is what is happening on this page

Reply:
![Screenshot 2020-04-07 at 16 26 27](https://user-images.githubusercontent.com/13540819/78688094-cc189700-78ec-11ea-91a6-53de52d266f6.png)

On Initial load, OnInit in pagination fires before OnInit in trainee-list.

This means is I hard code variables in url and hard refresh pagination control sends action first 
`if (params.pageIndex) {
      this.store.dispatch(new PaginateTrainees(params.pageIndex));
    } else {
      this.store.dispatch(new ResetTraineesPaginator());
    }`
then trainee-list sends its own OnInit
`if (params.active && params.direction) {
      this.store.dispatch(new SortTrainees(params.active, params.direction));
    } else {
      this.store.dispatch(new ResetTraineesSort());
    }

this.store.dispatch(new GetTrainees());
`

Reply:
![Screenshot 2020-04-07 at 16 27 51](https://user-images.githubusercontent.com/13540819/78688907-ab9d0c80-78ed-11ea-885a-f94f0350fcb1.png)